src/goto-symex/shadow_memory.cpp

```diff
@@ -95,12 +248,323 @@ void shadow_memoryt::symex_set_field(
   // To be implemented
 }
 
+// TODO: doxygen?
+static value_set_dereferencet::valuet get_shadow(
+  const exprt &shadow,
+  const object_descriptor_exprt &matched_base_descriptor,
+  const exprt &expr,
+  const namespacet &ns,
+  const messaget &log)
+{
+  object_descriptor_exprt shadowed_object = matched_base_descriptor;
+  shadowed_object.object() = shadow;
+  value_set_dereferencet::valuet shadow_dereference =
+    value_set_dereferencet::build_reference_to(shadowed_object, expr, ns);
+#ifdef DEBUG_SM
+  log.debug() << "Shadow memory: shadow-deref: "
+              << format(shadow_dereference.value) << messaget::eom;
+#endif
+  return shadow_dereference;
+}
+
+// TODO: doxygen?
+// foreach shadowed_address in SM:
+//   if(incompatible(shadow.object, object)) continue; // Type incompatibility
+//   base_match = object == shadow_object; // Do the base obj match the SM obj
+//   if(!base_match) continue;
+//   if(object is null) continue; // Done in the caller
+//   if(type_of(dereference object is void)
+//     // we terminate as we don't know how big is the memory at that location
+//   shadowed_dereference.pointer = deref(shadow.shadowed_object, expr);
+//   aggregate the object depending on the tyep
+//   expr_match = shadowed_dereference == expr
+//   if(!expr_match) continue;
+//   shadow_dereference = deref(shadow.object, expr);
+//   if(expr_match)
+//     result = shadow_dereference.value [exact match]
+//     break;
+//   if(base_match) [shadow memory matches]
+//     result += (expr_match,  shadow_dereference.value)
+//     break;
+//   result += (base_match & expr_match,  shadow_dereference.value)
+static std::vector<std::pair<exprt, exprt>> get_field(
```

Comment:
The naming of the function can be improved to `get_shadow_dereference_candidates` or similar.